### PR TITLE
Bugfix: make kernel generator wait for events on matrices

### PR DESCRIPTION
**Pull request by t4c1**
## Summary
Fixes kernel generator so it only executes kernels after appropriate events on matrices used by a kernel are complete.

## Tests
Added tests that check events are waited on.

## Side Effects

None.

## Checklist

- [x] Math issue #1342 

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by SteveBronder** [comment]
@serban-nicusor-toptal  Are all the windows boxes down or is there one still running so it's going slow? Looks like most tests are stopping while waiting for upstream windows boxes here

https://jenkins.mc-stan.org/blue/organizations/jenkins/CmdStan/detail/downstream_tests/1534/pipeline

**Comment by serban-nicusor-toptal** [comment]
There are two in use, one is down and can't be handled on-site because of covid-19.
On-Demand machines have a bug, I'm currently working on a fix with @rok-cesnovar to enable them back and get rid of the pressure.

**Comment by SteveBronder** [comment]
Sounds good thanks for the update!

**Comment by rok-cesnovar** [comment]
There was a huge backlog of Windows jobs. Its unfortunate that the Windows jobs take the longest and we also have the least workers (and obviousyl the least stable).

This job will finish in 20 minutes https://jenkins.mc-stan.org/blue/organizations/jenkins/Stan/detail/downstream_tests/1417/pipeline then this PR should be next.

File: stan/math/opencl/kernel_generator/operation_cl_lhs.hpp [hunk 1]
```diff
@@ -123,6 +124,20 @@ class operation_cl_lhs : public operation_cl<Derived, Scalar, Args...> {
...
+
+  /**
+   * Adds all read and write events on any matrices used by nested expressions
+   * to a list and clears them from those matrices.
+   * @param[out] events List of all events.
+   */
+  inline void get_clear_read_write_events(
+      std::vector<cl::Event>& events) const {
+    index_apply<N>([&](auto... Is) {
+      (void)std::initializer_list<int>{
```
Comment (SteveBronder) [inline stan/math/opencl/kernel_generator/operation_cl_lhs.hpp]: Should this be `static_cast<void>()`? I'm surprised this is getting through the cpplint

Comment (t4c1) [reply stan/math/opencl/kernel_generator/operation_cl_lhs.hpp]: Yeah. Fixed.

**Pull request merged**